src/app/src/main/kotlin/org/vaccineimpact/api/app/logic/CoverageLogic.kt

```diff
@@ -82,13 +120,65 @@ class RepositoriesCoverageLogic(private val modellingGroupRepository: ModellingG
                 )
             }.asSequence()
         }.toList()
+
+        val setsWithMissingRows = expectedRowLookup.missingRows()
+        if (setsWithMissingRows.any())
+        {
+            throw MissingRowsError(rowErrorMessage(setsWithMissingRows))
+        }
         touchstoneRepository.saveCoverageForTouchstone(touchstoneVersionId, records)
     }
 
-    constructor(repositories: Repositories) : this(repositories.modellingGroup,
-            repositories.responsibilities,
-            repositories.touchstone,
-            repositories.scenario)
+    private fun validate(vaccine: String,
+                         row: CoverageIngestionRow,
+                         countries: List<String>,
+                         expectedRowLookup: CoverageRowLookup): CoverageRowLookup
+    {
+        if (!expectedGAVICoverageYears.contains(row.year))
```

Comment:
This is getting called for both routine and campaign isn't it? I thought the expected years only applied to routine (I'm not really sure what the rules are for campaigns but the ticket says "campaigns can be any number of years") Is the expectation that campaign years can only be subset of routine years (won't accept any projected campaigns after 2030).

Reply:
Yeah, that was my understanding, but I will double check...

Reply:
Ok this seems like a correct assumption.